File: tools/convert_blog_post/src/anytype_simplied/workspace.rs
use std::collections::BTreeMap;

use crate::{
    anytype::object::AnytypeObject,
    anytype_proto::{
        anytype::SnapshotWithType,
        trait_impl::{get_field_value, to_val_map},
    },
};

use anyhow::Error;
use serde::{Deserialize, Serialize};

use super::{
    attributes::AttributeMap,
    collection::CollectionMap,
    page::PageMap,
    user::{User, UserMap},
};

pub type WorkspaceId = String;
pub type WorkspaceMap = BTreeMap<WorkspaceId, Workspace>;

#[derive(Default, Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct Workspace {
    pub id: WorkspaceId,
    pub name: String,
    pub description: String,
    pub projects: PageMap,
    pub collections: CollectionMap,
    pub users: UserMap,
    // pub tags: BTreeMap<&'static str, Tag>,
    #[serde(skip_serializing)]
    pub attributes: AttributeMap,
    // pub pages: HashMap<String, Page>,
    // pub relations: HashMap<String, Relation>,
}

impl Workspace {
    #[deprecated]
    pub fn from_anytype(raw: &AnytypeObject) -> Self {
        let data = raw.snapshot.data.details.clone();

        Self {
            id: data.id.clone(),
            name: data.name.unwrap_or_default().clone(),
            ..Default::default()
        }
    }

    pub fn from_anytype_proto(raw: &SnapshotWithType) -> Result<Self, Error> {
        let mut tmp = Self::default();

        let ref data_map = raw
            .snapshot
            .as_ref()
            .unwrap()
            .data
            .as_ref()
            .unwrap()
            .details
            .as_ref()
            .unwrap();
        // let data_map = snapshot.data.as_ref().unwrap().details.unwrap().fields;
        // // prost::Message::

        tmp.id = get_field_value(data_map, "id")?;
        tmp.name = get_field_value(data_map, "name")?;
        // tmp.description = get_field_value(data_map, "description")?;
        tmp.attributes = to_val_map(data_map)?;

        Ok(tmp)
    }

    pub fn add_user(&mut self, user: &User) {
        self.users.insert(user.id.clone(), user.clone());
    }
}

File: tools/convert_blog_post_ts/workspace.ts
import { SmartBlockType } from "../../protos/anytype/models";
import type { SnapshotWithType } from "../../protos/anytype/snapshot";
import type { AttributeMap } from "./attribute";
import type { CollectionMap } from "./collection";
import { getFieldValue } from "./common";
import type { PageMap } from "./page";
import type { Tag } from "./tag";
import type { User, UserMap } from "./user";

export type WorkspaceId = string;
export type WorkspaceMap = Map<WorkspaceId, Workspace>;

export interface Workspace {
  id: WorkspaceId;
  name: string;
  description: string;
  projects: PageMap;
  collections: CollectionMap;
  users: UserMap;
  tags: Map<string, Tag>;
  attributes: AttributeMap;
  // pub pages: HashMap<string, Page>,
  // pub relations: HashMap<string, Relation>,
}

export const getWorkspaces = (
  snapshot_list: SnapshotWithType[],
): Workspace[] => {
  return snapshot_list
    .filter((snapshot) => snapshot.sbType === SmartBlockType.Workspace)
    .map(fromAnytype);
};

export function fromAnytype(raw: SnapshotWithType): Workspace {
  const tmp: Workspace = {
    id: "",
    name: "",
    description: "",
    projects: new Map(),
    collections: new Map(),
    users: new Map(),
    tags: new Map(),
    attributes: new Map(),
  };
  const dataMap = raw.snapshot?.data?.details;

  if (!dataMap) {
    throw new Error("Data map is undefined");
  }
  // tmp.id = get_field_value(data_map, "id");
  tmp.id = getFieldValue(dataMap, "id");
  tmp.name = getFieldValue(dataMap, "name");
  // tmp.description = get_field_value(data_map, "description")?;
  // tmp.attributes = getFieldValue(dataMap, "attributes");
  // dataMap.get("attributes")?.value?.map_value?.fields ?? new Map();

  return tmp;
}

export function addUser(self: Workspace, user: User) {
  // self.users.insert(user.id.clone(), user.clone());
  self.users.set(user.id, user);
}
